### PR TITLE
Feat/hvb rwa keepers

### PR DESCRIPTION
**Pull request by dbkbali**
- RwaJarJobFactory and ScheduledJarVoidJob have been completed
- Next step is to review with @hexonaut and get feedback on the proposed structure and approach
- If he likes the approach - I will complete similar jobs for the RwaUrn wipe
Note: will also refactor the test setup accordingly once we have received feedback

### CODE REVIEW

File: src/tests/ScheduledJarVoidJob.t.sol [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        mcd.dai().transfer(rwaJars[RWAA],jarJob.THRESHOLD());
+        assertEq(mcd.dai().balanceOf(address(jarJob.rwaJar())), jarJob.THRESHOLD(), "Balance of RwaJar is not zero");
+        
+        (canWork, args) = jarJob.workable(NET_A);
+        assertTrue(!canWork, "Should not be able to work");
+
+    }
+
+    
+    function _createFakeDai(address usr, uint256 wad) private {
```
Comment (nduchak) [inline src/tests/ScheduledJarVoidJob.t.sol:228]: You can use `deal(address(dai), usr, wad)` instead,

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+  "solidity.compileUsingRemoteVersion": "v0.8.14+commit.80d49f37"
+}
```
Comment (amusingaxl) [inline .vscode/settings.json:3]: You probably want to leave editor-specific config out of version control.

File: RwaToken.json [hunk 3]
```diff
@@ -0,0 +1,3815 @@
+{
```
Comment (amusingaxl) [inline RwaToken.json:1]: Is this file still needed?

File: src-0_6_x/Contract6.sol [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (amusingaxl) [inline src-0_6_x/Contract6.sol:1]: You probably want to remove this file.

File: remappings.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+mip21-toolkit/=lib-0_6_x/mip21-toolkit/src
```
Comment (amusingaxl) [inline remappings.txt:1]: I was under the impression that Foundry would be able to find the proper files without requiring them to be explicitly added here. Maybe I was wrong.

File: src/RwaJarJobFactory.sol [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+ * @author David Krett <david@w2d.co>
+ * @notice Scans the RWA Registry for Urn and Jar jobs and creates jobs if they dont exist
+ */
+contract RwaJarJobFactory is TimedJob {
+    
+    using EnumerableSet for EnumerableSet.AddressSet;
+        
+    EnumerableSet.AddressSet private jobs;
+
+    uint256 private _lastRegistyIndex=0;
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:57]: Typo:
```suggestion
    uint256 private _lastRegistryIndex=0;
```

File: src/RwaJarJobFactory.sol [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+    enum DealStatus {
+        NONE, // The deal does not exist.
+        ACTIVE, // The deal is active.
+        FINALIZED // The deal was finalized.
+    }
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:63]: Why do you need to repeat this here instead of referencing it like `RWARegistryLike.DealStatus`?

File: src/RwaJarJobFactory.sol [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+    RWARegistryLike public immutable rwaRegistry;
+
+    mapping(bytes32 => address) public jarJobs;
+    mapping(address => bytes32 ) public jobIlks;
+
+    bytes32 constant public COMPONENT = "jar";
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:70]: I cannot think of a good reason why this should not be an internal constant. Nobody will ever need to "consume" this info.
```suggestion
    bytes32 constant internal COMPONENT = "jar";
```

File: src/RwaJarJobFactory.sol [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+    /**
+    * @notice The jar job identified by `ilk` was removed from the jobs network.
+    * @param jar The jar address.
+    */
+    event RemoveJarJob(address indexed jar);
+
+    constructor(address _sequencer, address _rwaRegistry, uint256 _duration) TimedJob(_sequencer, _duration) {
+        rwaRegistry = RWARegistryLike(_rwaRegistry);
+    }
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:86]: The order of the parameters is weird here. Maybe consider moving the registry to the end of the list:
```suggestion
    constructor(address _sequencer, uint256 _duration, address _rwaRegistry) TimedJob(_sequencer, _duration) {
        rwaRegistry = RWARegistryLike(_rwaRegistry);
    }
```

File: src/RwaJarJobFactory.sol [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    */
+    function shouldUpdate() internal override view returns (bool) {
+        // If any of these returns true an update is needed.
+        if (_checkJarsWithoutJobs() == true) {
+            return true;
+        }
+        if (_checkActiveJobsForInactiveDeals() == true) {
+            return true;
+        }
+        // Note Edge case check for changed deals
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:100]: This comment does not tell much. Would you mind expanding on it?

File: src/RwaJarJobFactory.sol [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+    /**
+    * @notice checks the RWA registry for new or redundant jar jobs.
+    */
+    function update() internal override {
+        // store the last index of registy jobs 
+        // add in new and remove any inactive jobs
+        // create jobs for new jars
+        _checkForNewJarJobs();
+        _checkForTerminatedJobs();
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:112]: The `check*` prefix is being used in the read-only part of the job. Since we're actually doing something, calling it "check***" sounds a bit strange. I'd change the names of these methods to something like:
```suggestion
        _addNewJobs();
        _removeTerminatedJobs();
```

File: src/RwaJarJobFactory.sol [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+    }
+
+    /**
+     * @notice iterates over RwaRegistry and returns true if jobs need to be deployed
+    */
+    function _checkJarsWithoutJobs() internal view returns (bool) {
+        bytes32[] memory ilks = rwaRegistry.iter(_lastRegistyIndex, rwaRegistry.count());
+        for (uint256 i = 0; i < ilks.length; i++) {
+            try rwaRegistry.ilkToDeal(ilks[i]) returns (RWARegistryLike.DealStatus status, uint248) {
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:122]: `ilkToDeal` is a mapping getter, which means it will never fail, so the `try..catch` here is superfluous.

File: src/RwaJarJobFactory.sol [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+            if (status == RWARegistryLike.DealStatus.ACTIVE) {
+                try rwaRegistry.getComponent(ilks[i], COMPONENT) returns (address addr, uint88) {
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:124]: If you are not using the second component of the tuple, you can omit its type also:
```suggestion
                try rwaRegistry.getComponent(ilks[i], COMPONENT) returns (address addr,) {
```

File: src/RwaJarJobFactory.sol [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+        return false;
+    }
+
+    /**
+     * @notice iterates over RwaRegistry and deploys a job for any active deals that have no job
+    */
+    function _checkForNewJarJobs() internal {
+        bytes32[] memory ilks = rwaRegistry.iter(_lastRegistyIndex, rwaRegistry.count());
+        for (uint256 i = 0; i < ilks.length; i++) {
+            try rwaRegistry.ilkToDeal(ilks[i]) returns (RWARegistryLike.DealStatus status, uint248) {
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:147]: Like above, `ilkToDeal` will never revert, so you can get rid of the outer `try..catch`.

Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:147]:
```suggestion
            (RWARegistryLike.DealStatus status, ) = rwaRegistry.ilkToDeal(ilks[i]);
```

File: src/RwaJarJobFactory.sol [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+    }
+
+    /**
+     * @notice creates a new jar job for the given jar address
+     * @param jar The jar address.
+    */
+    function _createJarJob(address jar) internal returns (address){
+        ScheduledJarVoidJob jarJob = new ScheduledJarVoidJob(address(sequencer), jar, 15 days);
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:176]: `15 days` is a magic number here.
This should either be extracted into a constant or be an immutable parameter to be set by the constructor.

File: src/RwaJarJobFactory.sol [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    /**
+     * @notice iterates over active jobs and returns true if jobs need to be removed due to inactive jars
+     */
+    function _checkActiveJobsForInactiveDeals() internal view returns (bool) {        
+        address[] memory activeJobs = jobs.values();
+        for (uint i = 0; i < activeJobs.length; i++) {
+            if (activeJobs[i] != address(0)) {
+                try rwaRegistry.ilkToDeal(jobIlks[activeJobs[i]]) returns (RWARegistryLike.DealStatus status, uint248) {
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:188]: `ilkToDeal` never reverts.

File: src/RwaJarJobFactory.sol [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @notice iterates over active jobs and removes any jobs associated with inactive jars
+     */
+    function _checkForTerminatedJobs() internal {
+        address[] memory activeJobs = jobs.values();
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:204]: `jobs.values()` copies the entire `activeJobs` value array to memory. This can be expensive depending on the amount of jobs being created.

`EnumerableSet.AddressSet` has a [`length()`](https://github.com/makerdao/dss-cron/blob/8eac401daeea3f7546785e3a232f3c124b69a17c/src/utils/EnumerableSet.sol#L248-L250) and an [`at(uint)`](https://github.com/makerdao/dss-cron/blob/8eac401daeea3f7546785e3a232f3c124b69a17c/src/utils/EnumerableSet.sol#L262-L264) methods that allow you to iterate over it without copying.

File: src/RwaJarJobFactory.sol [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+        for (uint i = 0; i < activeJobs.length; i++) {
+            if (activeJobs[i] != address(0)) {
+                try rwaRegistry.ilkToDeal(jobIlks[activeJobs[i]]) returns (RWARegistryLike.DealStatus status, uint248) {
```
Comment (amusingaxl) [inline src/RwaJarJobFactory.sol:207]: `ilkToDealI()` never reverts.

File: src/ScheduledJarVoidJob.sol [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+ */
+contract ScheduledJarVoidJob is TimedJob {
+    
+    RwaJarLike public immutable rwaJar;
+
+    DaiAbstract public immutable dai;
+
+    uint256 public constant THRESHOLD = 500 * (10 ** 18);
+
+    constructor(address _sequencer, address _jar, uint256 _duration) TimedJob(_sequencer, _duration) {
```
Comment (amusingaxl) [inline src/ScheduledJarVoidJob.sol:45]: Order of arguments is a bit strange:
```suggestion
    constructor(address _sequencer, uint256 _duration, address _jar) TimedJob(_sequencer, _duration) {
```

File: src/base/TimedJob.sol [hunk 20]
```diff
@@ -45,6 +54,7 @@ abstract contract TimedJob is IJob {
         if (!sequencer.isMaster(network)) revert NotMaster(network);
         uint256 expiry = last + maxDuration;
         if (block.timestamp < expiry) revert TimerNotElapsed(block.timestamp, expiry);
+        if (!shouldUpdate()) revert ShouldUpdateIsFalse();
```
Comment (amusingaxl) [inline src/base/TimedJob.sol:57]: I don't think we should add this here, because `shouldUpdate()` can be very gas intensive.
The whole reason it exists in the `view` side is to save gas on execution.

File: src/RwaJarJobFactory.sol (hunk 9 above)
Comment (dbkbali) [reply src/RwaJarJobFactory.sol:86]: I am going to leave this - as it is consistent with other timed jobs (ie lerpJob). To me it intuitively makes sense - addresses first and putting the duration last ..

File: src/RwaJarJobFactory.sol (hunk 10 above)
Comment (dbkbali) [reply src/RwaJarJobFactory.sol:100]: Ok we can detect terminated deals by change in deal status, and we can check for new deals - but what we cannot check for a change in deal components - ie. we use a jar initially and then decide to remove it - As we dont check for this and to check only deal status - this is the edge case that I think we can agree to manually remove jobs if we were ever to do this.

File: src/RwaJarJobFactory.sol (hunk 11 above)
Comment (dbkbali) [reply src/RwaJarJobFactory.sol:112]: great names!

File: src/RwaJarJobFactory.sol (hunk 13 above)
Comment (dbkbali) [reply src/RwaJarJobFactory.sol:124]: 🙏🏻

File: src/RwaJarJobFactory.sol (hunk 17 above)
Comment (dbkbali) [reply src/RwaJarJobFactory.sol:204]: Curious what the incremental impact on gas costs is ie.. which is highly dependent on the expected number of jobs ... but will take your approach.

File: src/tests/RwaJarJobFactory.t.sol [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+        
+        // deploy components and register addresses
+        address[] memory addrs = new address[](3);
+        addrs[0] =  deployUrn(rwaIlks[ilkIndex]);
+        addrs[1] = mcd.chainlog().getAddress("MIP21_LIQUIDATION_ORACLE");
+        addrs[2] = deployJar(rwaIlks[ilkIndex]);
+        rwaRegistry.add(rwaIlks[ilkIndex], _names, addrs, variants);
+    }
+
+    function _addAnotherJar(uint256 ilkIndex) internal {
```
Comment (amusingaxl) [inline src/tests/RwaJarJobFactory.t.sol:202]: There should be a well-defined naming convention for internal methods. Some of them have the `_` prefix, some don't.

File: src/tests/RwaJarJobFactory.t.sol [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+        bytes32,
+        uint256,
+        string calldata,
+        uint48
+    ) external;
+    function tell(bytes32) external;
+    function cure(bytes32) external;
+    function cull(bytes32) external;
+    function good(bytes32) external view;
+}
```
Comment (amusingaxl) [inline src/tests/RwaJarJobFactory.t.sol:68]: I believe it would be better to move all this interface definitions to the end of the file.

File: src/tests/RwaJarJobFactory.t.sol [hunk 23]
```diff
@@ -0,0 +1,273 @@
...
+        assertEq(rwaRegistry.ilks(0), RWAA);
+    
+        // ilks RWAA is deployed with an urn and jar
+        bytes32[] memory components = rwaRegistry.listComponentNamesOf(RWAA);
+        assertEq(components.length, 3);
+        assertEq(components[0], _names[0 /*URN*/]);
+        assertEq(components[1], _names[1 /*LIQUIDATION_ORACLE*/]);
+        assertEq(components[2], _names[2 /*JAR*/]);
+    }
+}
```
Comment (amusingaxl) [inline src/tests/RwaJarJobFactory.t.sol:272]: Maybe move these tests before all the helper methods.d

File: src/tests/ScheduledJarVoidJob.t.sol [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+        address destination = address(this);
+        address rwaJoin = joinFab.newAuthGemJoin(mcdPauseProxy, ilk, rwaTokens[ilk]);
+        address temp = deployCode("out/RwaUrn2.sol/RwaUrn2.json", abi.encode(mcdVat, mcdJug, rwaJoin, daiJoin, destination));
+        RwaUrnLike(temp).rely(mcdPauseProxy);
+        RwaUrnLike(temp).deny(address(this));
+        rwaUrns[ilk] = temp;
+        return temp;
+    }
+
+    function _createVariants() internal pure returns (uint88[] memory) {
```
Comment (amusingaxl) [inline src/tests/ScheduledJarVoidJob.t.sol:168]: Here we have the same issue of some internal methods being prefixed with `_` and some not.

File: src/RwaJarJobFactory.sol (hunk 9 above)
Comment (amusingaxl) [reply src/RwaJarJobFactory.sol:86]: Then maybe move the other way around, passing the registry address first. The sequencer and the duration are more closely related (they are used in the `TimedJob` constructor).

File: src/RwaJarJobFactory.sol (hunk 10 above)
Comment (amusingaxl) [reply src/RwaJarJobFactory.sol:100]: If it's out of the scope of the job, maybe we should not even include this comment then, or just give the full context. It brings more confusion as it is IMO.

File: src/RwaJarJobFactory.sol (hunk 17 above)
Comment (amusingaxl) [reply src/RwaJarJobFactory.sol:204]: This is one of those things that we have little control over. Let's say we figure out a way to scale RWA deals. We can end up with dozens of active jobs at a time.

Maybe it's an early optimization, but it does not add much in terms of complexity of the code.

File: src/tests/ScheduledJarVoidJob.t.sol (hunk 1 above)
Comment (amusingaxl) [reply src/tests/ScheduledJarVoidJob.t.sol]: `deal()` will only affect ERC-20 Dai. `_createFakeDai()` also takes care of the internal state in the `vat`.

File: src/base/TimedJob.sol (hunk 20 above)
Comment (dbkbali) [reply src/base/TimedJob.sol:57]: This I need to think about as the problem is Sam uses this with other jobs - so I think we should avoid changing the underlying contracts....

File: src/tests/RwaJarJobFactory.t.sol (hunk 22 above)
Comment (dbkbali) [reply src/tests/RwaJarJobFactory.t.sol:68]: better for who? if I am reading the tests as my documentation I want to see the interface dependencies so I understand what is available - so maybe depends on perspective ...

File: src/tests/RwaJarJobFactory.t.sol (hunk 23 above)
Comment (dbkbali) [reply src/tests/RwaJarJobFactory.t.sol:272]: yes will try and organize a little more logically

File: src/base/TimedJob.sol (hunk 20 above)
[the remark this answers is not in the document]
Comment (dbkbali) [reply src/base/TimedJob.sol:57]: Yes lets talk about this as this contract is the base for other jobs - so dont want to modify this as it may break some of sams logic - we should probably make our own base contract if this is important .    also keep in mind that in the context of a multimillion dollar deal that the gas costs are likely marginal

File: src/RwaJarJobFactory.sol (hunk 15 above)
Comment (dbkbali) [reply src/RwaJarJobFactory.sol:157]: moved to constuctor
